### PR TITLE
Adding Validation Accuracy and Loss to C-API of Activity Classifier

### LINKED ISSUE #1641: C-API activity classifier training should support validation-set metrics
`activity_classifier::train` currently accepts an argument `validation_data`, but it's ignored. For example, in the Python version we can do:
```
>>> model = tc.activity_classifier.create(data1, validation_set=data2, session_id='session_id', target='target', prediction_window=50, max_iterations=10)
Pre-processing 500000 samples...
Using sequences of size 1000 for model creation.
Processed a total of 43 sessions.
Using GPU to create model (AMD Radeon Pro Vega 56)
+---------------------+---------------------+---------------------+---------------------+---------------------+---------------------+
| Iteration           | Train Accuracy      | Train Loss          | Validation Accuracy | Validation Loss     | Elapsed Time        |
+---------------------+---------------------+---------------------+---------------------+---------------------+---------------------+
| 1                   | 0.604               | 1.069               | 0.693               | 1.274               | 1.4                 | 
| 2                   | 0.714               | 0.730               | 0.801               | 0.780               | 2.9                 | 
| 3                   | 0.823               | 0.543               | 0.788               | 0.538               | 4.4                 | 
| 4                   | 0.845               | 0.438               | 0.871               | 0.366               | 5.9                 | 
| 5                   | 0.866               | 0.376               | 0.886               | 0.318               | 7.4                 | 
| 6                   | 0.888               | 0.315               | 0.893               | 0.294               | 8.8                 | 
| 7                   | 0.897               | 0.281               | 0.899               | 0.271               | 10.3                | 
| 8                   | 0.902               | 0.254               | 0.906               | 0.264               | 11.8                | 
| 9                   | 0.911               | 0.237               | 0.906               | 0.252               | 13.3                | 
| 10                  | 0.915               | 0.220               | 0.888               | 0.289               | 14.8                | 
+---------------------+---------------------+---------------------+---------------------+---------------------+---------------------+
Training complete
Total Time Spent: 14.7509s
```

### PR DESCRIPTION
**Pull request by shreyajain17**
fixes #1641

### CODE REVIEW

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 1]
```diff
@@ -106,17 +154,30 @@ void activity_classifier::train(gl_sframe data, std::string target_column_name,
...
 
-  // Instantiate the training dependencies: data iterator, compute context,
-  // backend NN model.
+  if (variant_is<gl_sframe>(validation_data) ){
+
+    training_table_printer_.reset(new table_printer(
+    { {"Iteration", 12}, {"Train Accuracy", 12}, {"Train Loss", 12}, {"Validation Accuracy", 12}, {"Validation Loss", 12}, {"Elapsed Time", 12} }));
+
+  }
+  // else if (variant_is<flex_string>(validation_data) ){
```
Comment (TobyRoseman) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Let's remove commented out code.

Comment (shreyajain17) [reply src/unity/toolkits/activity_classification/activity_classifier.cpp]: Yeah I am still working on this PR. I had uploaded it so that Nick can look at it because I was getting NAN values for accuracy and loss. I was not resetting the validation data iterator. Although that is fixed now but this code needs more work. 

The reason for keeping this code was that I want to eventually accept "auto" string as well. Then elaborate on splitting the dataset itself.

I will add you for request as soon as I am done.

**Comment by shreyajain17** [comment]
> Would it be reasonably possible to write a unit test for this?

I am not sure if there are any unit tests for the C++ implementation of activity classifier.

**Comment by nickjong** [comment]
It was on my to-do list for this sprint to add some of the infrastructure needed to make these tests less painful to write (I didn't think it was fair to ask Shreya to write this). It's something I'm currently planning to do for the next sprint. It's definitely something we need to have before we actually replace the shipping (Python-based) toolkit with the new C++ version.

File: src/unity/toolkits/activity_classification/activity_classifier.cpp (hunk 1 above)
Comment (TobyRoseman) [reply src/unity/toolkits/activity_classification/activity_classifier.cpp]: Sorry @shreyajain17 - I didn't realize it was WIP. I'll make sure to check for that in the future.

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 2]
```diff
@@ -152,18 +154,25 @@ void activity_classifier::train(gl_sframe data, std::string target_column_name,
 {
   // Begin printing progress.
   // TODO: Make progress printing optional.
-  
-  // TODO: Support validation set.
-  training_table_printer_.reset(new table_printer(
+
+  if (variant_is<gl_sframe>(validation_data) ){
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: It's looking like it could make sense to move the initialization of `training_table_printer_` into `init_train`, which is where we print the table header anyway.

(Also, let's make sure to run clang-format on this PR's commits too. We should aim for more consistent formatting than e.g. the parens spacing in this line.)

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 3]
```diff
@@ -152,18 +154,25 @@ void activity_classifier::train(gl_sframe data, std::string target_column_name,
...
-  
-  // Instantiate the training dependencies: data iterator, compute context,
-  // backend NN model.
+  }
+
   init_train(std::move(data), std::move(target_column_name),
              std::move(session_id_column_name), std::move(validation_data),
              std::move(opts));
 
+  // Instantiate the training dependencies: data iterator, compute context,
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Why move this comment down here? It's meant to document the `init_train` invocation

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 4]
```diff
@@ -507,18 +529,87 @@ void activity_classifier::perform_training_iteration() {
...
   // Report progress if we have an active table printer.
-  // TODO: Report validation metrics.
-  
-  if (training_table_printer_) {
-    training_table_printer_->print_progress_row(
-        iteration_idx, iteration_idx + 1, average_batch_accuracy, average_batch_loss, progress_time());
+  if (validation_data_iterator_ != nullptr) {
+
+    float cumulative_val_accuracy = 0.f;
+    float cumulative_val_loss = 0.f;
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Could we factor out the computation of `cumulative_val_accuracy` and `cumulative_val_loss` into a (single) separate (private) helper method?

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 5]
```diff
@@ -507,18 +529,87 @@ void activity_classifier::perform_training_iteration() {
...
+      std::map<std::string, shared_float_array> results =
+          training_model_->predict({
+              {"input", val.features},
+              {"labels", val.labels},
+              {"weights", val.weights},
+          });
+
+      const shared_float_array &val_output = results.at("output");
+
+      shared_float_array loss = results.at("loss");
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Nit: `val_output` and `loss` should logically have the same type, either both a const reference or both just a `shared_float_array` value. May as well have them both be references.

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 6]
```diff
@@ -507,18 +529,87 @@ void activity_classifier::perform_training_iteration() {
...
+        data_iterator::batch::chunk_info info = val.batch_info[i];
+        size_t num_predictions =
+            (info.num_samples + prediction_window - 1) / prediction_window;
+        size_t num_correct_predictions = count_correct_predictions(
+            num_classes, output_chunk, label_chunk, num_predictions);
+
+        cumulative_per_seq_accuracy +=
+            static_cast<float>(num_correct_predictions) / num_predictions;
+      }
+      cumulative_val_accuracy += cumulative_per_seq_accuracy;
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: I don't think we need both `cumulative_val_accuracy` and `cumulative_seq_accuracy`. Can just add each sequence's accuracy directly into `cumulative_val_accuracy`

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 7]
```diff
@@ -507,18 +529,87 @@ void activity_classifier::perform_training_iteration() {
...
+    }
+
   }
 
-  add_or_update_state({
-      { "training_iterations", iteration_idx + 1 },
-      { "training_accuracy" , average_batch_accuracy },
-      { "training_log_loss", average_batch_loss },
-  });
+  else {
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Right now the code is organized like:
```
if has_validation_data:
  compute validation metrics
  update progress table
  update state
else:
  update progress table
  update state
```

Could we instead do something more like:
```
  compute validation metrics if necessary
  update progress table
  update state
```
Where each step's behavior depends on whether validation data is available? To me, validation data availability should be the "lower order bit" from the point of view of code organization

**Comment by shreyajain17** [comment]
I was thinking about it. I think I am not handling the case if the given Frame for validation is empty and also I don't know what it does if the data is less then one batch size.

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 8]
```diff
@@ -106,7 +107,7 @@ float get_accuracy_per_batch(size_t prediction_window, size_t num_classes,
     
     cumulative_per_batch_accuracy += static_cast<float>(num_correct_predictions) / num_predictions;
   }
-  return cumulative_per_batch_accuracy / batch.batch_info.size();
+  return cumulative_per_batch_accuracy;
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Let's rename this helper function, since you've changed the semantics of the return value. 

I honestly don't know what this should really be called? Cumulative chunk accuracy?

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 9]
```diff
@@ -145,19 +146,49 @@ void activity_classifier::init_options(
   add_or_update_state(flexmap_to_varmap(options.current_option_values()));
 }
 
+std::tuple<float, float> activity_classifier::compute_validation_metrics(
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Let's at least add a quick comment documenting what the semantics of the two tuple values are (accuracy and loss)

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 10]
```diff
@@ -145,19 +146,49 @@ void activity_classifier::init_options(
...
+    data_iterator *validation_data_iterator_, size_t prediction_window,
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: I don't think we need to pass the `data_iterator*` since we can just use the class data member, right?

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 11]
```diff
@@ -145,19 +146,49 @@ void activity_classifier::init_options(
...
+        validation_data_iterator_->next_batch(batch_size);
+    std::map<std::string, shared_float_array> results =
+        training_model_->predict({
+            {"input", val.features},
+            {"labels", val.labels},
+            {"weights", val.weights},
+        });
+
+    const shared_float_array &output = results.at("output");
+    shared_float_array loss = results.at("loss");
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: I feel like `output` and `loss` should either both be values (`shared_float_array`) or references (`const shared_float_array&`)

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 12]
```diff
@@ -447,6 +496,19 @@ void activity_classifier::init_train(
...
+    data_iterator::parameters validation_data_params;
+    validation_data_params.data = validation_sf;
+    validation_data_params.target_column_name = target_column_name;
+    validation_data_params.session_id_column_name = session_id_column_name;
+    validation_data_params.feature_column_names = feature_column_names;
+    validation_data_params.prediction_window =
+        read_state<flex_int>("prediction_window");
+    validation_data_params.predictions_in_chunk = NUM_PREDICTIONS_PER_CHUNK;
+    validation_data_iterator_ = create_iterator(validation_data_params);
+  }
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: Don't forget to add the else case here, to guard against multiple calls to `init_train`

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 13]
```diff
@@ -539,33 +601,57 @@ void activity_classifier::perform_training_iteration() {
...
   float average_batch_accuracy = cumulative_batch_accuracy / num_batches;
 
-  // Report progress if we have an active table printer.
-  // TODO: Report validation metrics.
-  
-  if (training_table_printer_) {
-    training_table_printer_->print_progress_row(
-        iteration_idx, iteration_idx + 1, average_batch_accuracy, average_batch_loss, progress_time());
 
+  if (validation_data_iterator_ == nullptr) {
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: The remainder of this function is still hard to follow for me

Could we do something more like:
```
float average_val_accuracy;
float average_val_loss;
if (validation_data_iterator) {
  // Compute average_val_accuracy and average_val_loss
}

std::map<std::string, variant_type> state_updates({
  // training metrics
});
if (validation_data_iterator) {
  state_updates["validation_accuracy"] = ...;
  ...
}

if (training_table_printer_) {
  if (validation_data_iterator) {
    // Print six columns
  } else {
    // Print four columns
  }
}
```

(FWIW, I think doing everything with a single unified `if (validation_data_iterator_)` statement could make sense, but only if we refactor things to the point where the entire if-else statement fits on "one screen", without scrolling)

File: src/unity/toolkits/activity_classification/activity_classifier.cpp [hunk 14]
```diff
@@ -447,6 +502,21 @@ void activity_classifier::init_train(
...
+    gl_sframe validation_sf = variant_get_value<gl_sframe>(validation_data);
+    data_iterator::parameters validation_data_params;
+    validation_data_params.data = validation_sf;
+    validation_data_params.target_column_name = target_column_name;
+    validation_data_params.session_id_column_name = session_id_column_name;
+    validation_data_params.feature_column_names = feature_column_names;
+    validation_data_params.prediction_window =
+        read_state<flex_int>("prediction_window");
+    validation_data_params.predictions_in_chunk = NUM_PREDICTIONS_PER_CHUNK;
+    validation_data_iterator_ = create_iterator(validation_data_params);
```
Comment (nickjong) [inline src/unity/toolkits/activity_classification/activity_classifier.cpp]: It would be nice to change create_iterator to accept just the SFrame, since all the other `data_iterator::parameters` are the same every time. We can do this in the next PR, though

**Comment by nickjong** [comment]
Don't forget to remove the "WIP" prefix before merging!
